src/solve/moi.jl

```diff
@@ -4,20 +4,96 @@ const MOI = MathOptInterface
+function MOIOptimizationProblem(prob::OptimizationProblem)
+    num_cons = prob.ucons === nothing ? 0 : length(prob.ucons)
+    f = instantiate_function(prob.f, prob.u0, prob.f.adtype, prob.p, num_cons)
+    T = eltype(prob.u0)
+    n = length(prob.u0)
+    return MOIOptimizationProblem(
+        f,
+        prob.u0,
+        prob.p,
+        zeros(T, num_cons, n),
+        zeros(T, n, n),
+        Matrix{T}[zeros(T, n, n) for i in 1:num_cons],
+    )
+end
+
+MOI.features_available(::MOIOptimizationProblem) = [:Grad, :Hess, :Jac]
+
+function MOI.initialize(
+    moiproblem::MOIOptimizationProblem,
+    requested_features::Vector{Symbol},
+)
+    available_features = MOI.features_available(moiproblem)
+    for feat in requested_features
+        if !(feat in available_features)
+            error("Unsupported feature $feat")
+            # TODO: implement Jac-vec and Hess-vec products
+            # for solvers that need them
+        end
+    end
+    return
+end
+
+function MOI.eval_objective(moiproblem::MOIOptimizationProblem, x)
+    return moiproblem.f(x, moiproblem.p)
+end
 
-MOI.eval_constraint(moiproblem::MOIOptimizationProblem, g, x) = g .= moiproblem.f.cons(x)
+function MOI.eval_constraint(moiproblem::MOIOptimizationProblem, g, x)
+    g .= moiproblem.f.cons(x)
+    return
+end
 
-MOI.eval_objective_gradient(moiproblem::MOIOptimizationProblem, G, x) = moiproblem.f.grad(G, x)
+function MOI.eval_objective_gradient(moiproblem::MOIOptimizationProblem, G, x)
+    moiproblem.f.grad(G, x)
+    return
+end
 
-function MOI.eval_hessian_lagrangian(moiproblem::MOIOptimizationProblem{T}, h, x, σ, μ) where {T}
+# This structure assumes the calculation of moiproblem.J is dense.
+function MOI.jacobian_structure(moiproblem::MOIOptimizationProblem)
+    rows, cols = size(moiproblem.J)
+    return Tuple{Int,Int}[(i, j) for j in 1:cols for i in 1:rows]
+end
+
+function MOI.eval_constraint_jacobian(moiproblem::MOIOptimizationProblem, j, x)
+    if isempty(j)
+        return
+    elseif moiproblem.f.cons_j === nothing
+        error(
+            "Use OptimizationFunction to pass the derivatives or " *
+            "automatically generate them with one of the autodiff backends",
+        )
+    end
+    moiproblem.f.cons_j(moiproblem.J, x)
+    for i in eachindex(j)
+        j[i] = moiproblem.J[i]
+    end
+    return
+end
+
+# Because the Hessian is symmetrical, we choose to store the upper-triangular
+# component. We also assume that it is dense.
+function MOI.hessian_lagrangian_structure(moiproblem::MOIOptimizationProblem)
+    num_vars = length(moiproblem.u0)
+    return Tuple{Int,Int}[(row, col) for col in 1:num_vars for row in 1:col]
+end
+
+function MOI.eval_hessian_lagrangian(
+    moiproblem::MOIOptimizationProblem{T},
+    h,
+    x,
+    σ,
+    μ,
+) where {T}
     n = length(moiproblem.u0)
+    # TODO(odow): what is this? Why is it called?
     a = zeros(n, n)
     moiproblem.f.hess(a, x)
```

Comment:
Can anyone explain what this call is doing here? It looks like we just create an empty array and then compute the Hessian (of the objective?), only to throw it away.

Reply:
I agree, it looks redundant.

Reply:
Removed it.